src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1454,32 +1455,106 @@ let CheckMultipleInputsInParallel
+                List.toArray parallelFiles
+                |> ArrayParallel.map (fun (input, logger) ->
+                    cancellable {
+                        // this is taken mostly from CheckOneInputAux, the case where the impl has no signature file
+                        let file =
+                            match input with
+                            | ParsedInput.ImplFile file -> file
+                            | ParsedInput.SigFile _ -> failwith "not expecting a signature file for now"
+
+                        let tcSink = TcResultsSink.NoSink
+
+                        // Typecheck the implementation file
+                        let! topAttrs, implFile, tcEnvAtEnd, createsGeneratedProvidedTypes =
+                            CheckOneImplFile(
+                                tcGlobals,
+                                amap,
+                                sequentialTcState.tcsCcu,
+                                sequentialTcState.tcsImplicitOpenDeclarations,
+                                checkForErrors,
+                                conditionalDefines,
+                                TcResultsSink.NoSink,
+                                tcConfig.internalTestSpanStackReferring,
+                                sequentialTcState.tcsTcImplEnv,
+                                None,
+                                file
+                            )
+
+                        return
+                            (fun tcState ->
+                                let tcState =
+                                    { tcState with
+                                        tcsCreatesGeneratedProvidedTypes =
+                                            tcState.tcsCreatesGeneratedProvidedTypes || createsGeneratedProvidedTypes
+                                    }
+
+                                let ccuSigForFile, updateTcState =
+                                    AddCheckResultsToTcState
+                                        (tcGlobals,
+                                         amap,
+                                         false,
+                                         prefixPathOpt,
+                                         tcSink,
+                                         tcState.tcsTcImplEnv,
+                                         input.QualifiedName,
+                                         implFile.Signature)
+                                        tcState
+
+                                Choice1Of2(tcEnvAtEnd, topAttrs, Some implFile, ccuSigForFile), logger, updateTcState)
+                    }
+                    |> Cancellable.runWithoutCancellation)
+                |> fun results ->
+                    ((sequentialTcState, sequentialPriorErrors, ImmutableQueue.Empty), results)
+                    ||> Array.fold (fun (tcState, priorErrors, partialResults) result ->
+                        let partialResult, logger, nextTcState = result tcState
```

Comment:
@safesparrow here, I fold the callback over here and everything plays out as it would in sequence.